File: api/rest/handler/memory_test.go
package handler_test

import (
	"context"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"
	"time"

	"github.com/CoscaAI/cosca/api/rest/handler"
	"github.com/CoscaAI/cosca/internal/memory"
)

// newTestMemoryEngine creates a fully initialized memory engine backed by a
// temporary directory. AutoPrune is disabled to avoid goroutine leaks under
// -race, and SnapshotOnPromote is disabled to keep Promote fast.
func newTestMemoryEngine(t *testing.T) *memory.MemoryEngine {
	t.Helper()
	cfg := memory.EngineConfig{
		DataDir:           t.TempDir(),
		AutoPrune:         false,
		SnapshotOnPromote: false,
	}
	engine, err := memory.NewEngine(memory.WithConfig(cfg))
	if err != nil {
		t.Fatalf("failed to create memory engine: %v", err)
	}
	t.Cleanup(func() { _ = engine.Close() })
	return engine
}

// newTestMemoryHandler creates a MemoryHandler with a real engine.
func newTestMemoryHandler(t *testing.T) *handler.MemoryHandler {
	t.Helper()
	return handler.NewMemoryHandler(newTestMemoryEngine(t), nil)
}

// storeTestMemory stores a memory record through the handler and returns
// the parsed StoreResponse.
func storeTestMemory(t *testing.T, h *handler.MemoryHandler, body string) handler.StoreResponse {
	t.Helper()
	req := httptest.NewRequest("POST", "/v1/memory/store", strings.NewReader(body))
	req.Header.Set("Content-Type", "application/json")
	w := httptest.NewRecorder()
	h.Store(w, req)

	if w.Code != http.StatusCreated {
		t.Fatalf("expected 201, got %d: %s", w.Code, w.Body.String())
	}
	var resp handler.StoreResponse
	if err := json.Unmarshal(w.Body.Bytes(), &resp); err != nil {
		t.Fatalf("failed to unmarshal store response: %v", err)
	}
	if resp.ID == "" {
		t.Fatal("expected non-empty id in store response")
	}
	return resp
}

// TestMemoryStoreSuccess verifies Store returns 201 with an id and
// created_at timestamp.
func TestMemoryStoreSuccess(t *testing.T) {
	h := newTestMemoryHandler(t)

	body := `{"type":"fact","layer":"session","content":"Cosca was started in 2026","priority":5}`
	req := httptest.NewRequest("POST", "/v1/memory/store", strings.NewReader(body))
	req.Header.Set("Content-Type", "application/json")
	w := httptest.NewRecorder()
	h.Store(w, req)

	if w.Code != http.StatusCreated {
		t.Fatalf("expected 201, got %d: %s", w.Code, w.Body.String())
	}

	var resp handler.StoreResponse
	if err := json.Unmarshal(w.Body.Bytes(), &resp); err != nil {
		t.Fatalf("failed to unmarshal: %v", err)
	}
	if resp.ID == "" {
		t.Error("expected non-empty id")
	}
	if resp.CreatedAt == "" {
		t.Error("expected non-empty created_at")
	}
}

// TestMemoryStoreFullRequest verifies Store accepts TTL, metadata, and
// scope fields and still succeeds.
func TestMemoryStoreFullRequest(t *testing.T) {
	h := newTestMemoryHandler(t)

	body := `{"type":"bug","layer":"session","scope":"api/rest","content":"Handler bug found","priority":9,"ttl":"2h","metadata":{"source":"test"}}`
	req := httptest.NewRequest("POST", "/v1/memory/store", strings.NewReader(body))
	req.Header.Set("Content-Type", "application/json")
	w := httptest.NewRecorder()
	h.Store(w, req)

	if w.Code != http.StatusCreated {
		t.Fatalf("expected 201, got %d: %s", w.Code, w.Body.String())
	}
}

// TestMemoryStoreInvalidLayer verifies Store returns 500 (and does not panic)
// when the engine rejects the layer. The engine has no backing store for
// unknown layers, so Store returns (nil, err); the handler must not
// dereference the nil record when building the error-path audit event.
func TestMemoryStoreInvalidLayer(t *testing.T) {
	h := newTestMemoryHandler(t)

	body := `{"content":"x","layer":"bogus"}`
	req := httptest.NewRequest("POST", "/v1/memory/store", strings.NewReader(body))
	req.Header.Set("Content-Type", "application/json")
	w := httptest.NewRecorder()
	h.Store(w, req)

	if w.Code != http.StatusInternalServerError {
		t.Fatalf("expected 500, got %d: %s", w.Code, w.Body.String())
	}
	if !strings.Contains(w.Body.String(), "store failed") {
		t.Errorf("expected body mentioning 'store failed', got %q", w.Body.String())
	}
}

// TestMemorySearchSuccess verifies Search returns stored records and covers
// recordToAPI through the search response.
func TestMemorySearchSuccess(t *testing.T) {
	h := newTestMemoryHandler(t)

	storeTestMemory(t, h, `{"type":"fact","layer":"session","content":"Golang concurrency uses goroutines"}`)

	req := httptest.NewRequest("GET", "/v1/memory/search?query=Golang", nil)
	w := httptest.NewRecorder()
	h.Search(w, req)

	if w.Code != http.StatusOK {
		t.Fatalf("expected 200, got %d: %s", w.Code, w.Body.String())
	}

	var resp handler.MemorySearchResponse
	if err := json.Unmarshal(w.Body.Bytes(), &resp); err != nil {
		t.Fatalf("failed to unmarshal: %v", err)
	}
	if resp.Total != 1 {
		t.Errorf("expected 1 result, got %d", resp.Total)
	}
	if len(resp.Records) != 1 {
		t.Fatalf("expected 1 record, got %d", len(resp.Records))
	}
	if resp.Records[0].Content == "" {
		t.Error("expected record content to be populated (recordToAPI mapping)")
	}
	if resp.Records[0].Layer != "session" {
		t.Errorf("expected layer 'session', got %q", resp.Records[0].Layer)
	}
}

// TestMemorySearchNoResults verifies Search returns an empty records array
// when nothing matches.
func TestMemorySearchNoResults(t *testing.T) {
	h := newTestMemoryHandler(t)

	storeTestMemory(t, h, `{"type":"fact","layer":"session","content":"Golang concurrency uses goroutines"}`)

	req := httptest.NewRequest("GET", "/v1/memory/search?query=quantum-flux-compensator", nil)
	w := httptest.NewRecorder()
	h.Search(w, req)

	if w.Code != http.StatusOK {
		t.Fatalf("expected 200, got %d: %s", w.Code, w.Body.String())
	}

	var resp handler.MemorySearchResponse
	if err := json.Unmarshal(w.Body.Bytes(), &resp); err != nil {
		t.Fatalf("failed to unmarshal: %v", err)
	}
	if resp.Records == nil {
		t.Error("expected non-nil records array")
	}
	if resp.Total != 0 {
		t.Errorf("expected 0 total, got %d", resp.Total)
	}
}

// TestMemorySearchWithParams verifies Search honors types, layers, and
// limit query parameters.
func TestMemorySearchWithParams(t *testing.T) {
	h := newTestMemoryHandler(t)

	storeTestMemory(t, h, `{"type":"fact","layer":"session","content":"Alpha memory content"}`)
	storeTestMemory(t, h, `{"type":"bug","layer":"session","content":"Beta memory content"}`)

	req := httptest.NewRequest("GET", "/v1/memory/search?query=memory&types=fact&layers=session&limit=1", nil)
	w := httptest.NewRecorder()
	h.Search(w, req)

	if w.Code != http.StatusOK {
		t.Fatalf("expected 200, got %d: %s", w.Code, w.Body.String())
	}

	var resp handler.MemorySearchResponse
	if err := json.Unmarshal(w.Body.Bytes(), &resp); err != nil {
		t.Fatalf("failed to unmarshal: %v", err)
	}
	if resp.Total > 1 {
		t.Errorf("expected limit 1 to cap results, got %d", resp.Total)
	}
}

// TestMemorySearchEmptyQuery verifies Search works with an empty query
// (returns recent records).
func TestMemorySearchEmptyQuery(t *testing.T) {
	h := newTestMemoryHandler(t)

	storeTestMemory(t, h, `{"type":"fact","layer":"session","content":"Recent record content"}`)

	req := httptest.NewRequest("GET", "/v1/memory/search", nil)
	w := httptest.NewRecorder()
	h.Search(w, req)

	if w.Code != http.StatusOK {
		t.Fatalf("expected 200, got %d: %s", w.Code, w.Body.String())
	}
}

// TestMemoryGetSuccess verifies Get returns the stored record and covers
// recordToAPI through the get response.
func TestMemoryGetSuccess(t *testing.T) {
	h := newTestMemoryHandler(t)

	saved := storeTestMemory(t, h, `{"type":"decision","layer":"session","content":"Decision content"}`)

	req := httptest.NewRequest("GET", "/v1/memory/get?id="+saved.ID+"&layer=session", nil)
	w := httptest.NewRecorder()
	h.Get(w, req)

	if w.Code != http.StatusOK {
		t.Fatalf("expected 200, got %d: %s", w.Code, w.Body.String())
	}

	var resp handler.GetResponse
	if err := json.Unmarshal(w.Body.Bytes(), &resp); err != nil {
		t.Fatalf("failed to unmarshal: %v", err)
	}
	if resp.Record.ID != saved.ID {
		t.Errorf("expected id %q, got %q", saved.ID, resp.Record.ID)
	}
	if resp.Record.Type != "decision" {
		t.Errorf("expected type 'decision', got %q", resp.Record.Type)
	}
	if resp.Record.CreatedAt == "" {
		t.Error("expected created_at to be formatted (recordToAPI mapping)")
	}
}

// TestMemoryGetNotFound verifies Get returns 404 for an unknown id.
func TestMemoryGetNotFound(t *testing.T) {
	h := newTestMemoryHandler(t)

	req := httptest.NewRequest("GET", "/v1/memory/get?id=missing-id&layer=session", nil)
	w := httptest.NewRecorder()
	h.Get(w, req)

	if w.Code != http.StatusNotFound {
		t.Fatalf("expected 404, got %d: %s", w.Code, w.Body.String())
	}
}

// TestMemoryDeleteSuccess verifies Delete removes a stored record.
func TestMemoryDeleteSuccess(t *testing.T) {
	h := newTestMemoryHandler(t)

	saved := storeTestMemory(t, h, `{"type":"fact","layer":"session","content":"To be deleted"}`)

	req := httptest.NewRequest("DELETE", "/v1/memory/delete?id="+saved.ID+"&layer=session", nil)
	w := httptest.NewRecorder()
	h.Delete(w, req)

	if w.Code != http.StatusOK {
		t.Fatalf("expected 200, got %d: %s", w.Code, w.Body.String())
	}

	var resp map[string]string
	if err := json.Unmarshal(w.Body.Bytes(), &resp); err != nil {
		t.Fatalf("failed to unmarshal: %v", err)
	}
	if resp["status"] != "deleted" {
		t.Errorf("expected status 'deleted', got %q", resp["status"])
	}

	// The record must actually be gone.
	getReq := httptest.NewRequest("GET", "/v1/memory/get?id="+saved.ID+"&layer=session", nil)
	w2 := httptest.NewRecorder()
	h.Get(w2, getReq)
	if w2.Code != http.StatusNotFound {
		t.Errorf("expected 404 after delete, got %d", w2.Code)
	}
}

// TestMemoryDeleteNotFound verifies Delete returns 404 when the record does
// not exist.
func TestMemoryDeleteNotFound(t *testing.T) {
	h := newTestMemoryHandler(t)

	req := httptest.NewRequest("DELETE", "/v1/memory/delete?id=missing-id&layer=session", nil)
	w := httptest.NewRecorder()
	h.Delete(w, req)

	if w.Code != http.StatusNotFound {
		t.Fatalf("expected 404, got %d: %s", w.Code, w.Body.String())
	}
}

// TestMemoryPromoteSuccess verifies Promote moves a record from session to
// project and covers recordToAPI through the promote response.
func TestMemoryPromoteSuccess(t *testing.T) {
	h := newTestMemoryHandler(t)

	saved := storeTestMemory(t, h, `{"type":"pattern","layer":"session","content":"Promotable pattern"}`)

	body := `{"id":"` + saved.ID + `","from_layer":"session","to_layer":"project"}`
	req := httptest.NewRequest("POST", "/v1/memory/promote", strings.NewReader(body))
	req.Header.Set("Content-Type", "application/json")
	w := httptest.NewRecorder()
	h.Promote(w, req)

	if w.Code != http.StatusOK {
		t.Fatalf("expected 200, got %d: %s", w.Code, w.Body.String())
	}

	var resp handler.PromoteResponse
	if err := json.Unmarshal(w.Body.Bytes(), &resp); err != nil {
		t.Fatalf("failed to unmarshal: %v", err)
	}
	if resp.Record.ID != saved.ID {
		t.Errorf("expected id %q, got %q", saved.ID, resp.Record.ID)
	}
	if resp.Record.Layer != "project" {
		t.Errorf("expected promoted layer 'project', got %q", resp.Record.Layer)
	}
}

// TestMemoryPromoteInvalidLayer verifies Promote returns 500 when the target
// layer has no backing store.
func TestMemoryPromoteInvalidLayer(t *testing.T) {
	h := newTestMemoryHandler(t)

	saved := storeTestMemory(t, h, `{"type":"fact","layer":"session","content":"Promote to nowhere"}`)

	body := `{"id":"` + saved.ID + `","from_layer":"session","to_layer":"bogus"}`
	req := httptest.NewRequest("POST", "/v1/memory/promote", strings.NewReader(body))
	req.Header.Set("Content-Type", "application/json")
	w := httptest.NewRecorder()
	h.Promote(w, req)

	if w.Code != http.StatusInternalServerError {
		t.Fatalf("expected 500, got %d: %s", w.Code, w.Body.String())
	}
}

// TestMemoryStats verifies Stats returns a layers map even with a nil
// request body.
func TestMemoryStats(t *testing.T) {
	h := newTestMemoryHandler(t)

	storeTestMemory(t, h, `{"type":"fact","layer":"session","content":"Stats record"}`)

	req := httptest.NewRequest("GET", "/v1/memory/stats", nil)
	w := httptest.NewRecorder()
	h.Stats(w, req)

	if w.Code != http.StatusOK {
		t.Fatalf("expected 200, got %d: %s", w.Code, w.Body.String())
	}

	var resp handler.MemoryStatsResponse
	if err := json.Unmarshal(w.Body.Bytes(), &resp); err != nil {
		t.Fatalf("failed to unmarshal: %v", err)
	}
	if resp.Layers == nil {
		t.Fatal("expected non-nil layers map")
	}
	if len(resp.Layers) == 0 {
		t.Fatal("expected at least one layer in stats")
	}
	if stats, ok := resp.Layers["session"]; ok && stats.RecordCount < 1 {
		t.Errorf("expected at least 1 record in session layer, got %d", stats.RecordCount)
	}
}

// TestMemoryStoreRecordExpires ensures the engine TTL expiry path used by
// Retrieve (Get handler 404) does not leak goroutines — creates a record
// with a past CreatedAt so Retrieve reports it as expired.
func TestMemoryStoreRecordExpires(t *testing.T) {
	engine := newTestMemoryEngine(t)
	ctx := context.Background()

	expired := memory.MemoryRecord{
		ID:        "expired-id",
		Type:      memory.MemoryType("fact"),
		Layer:     memory.LayerSession,
		Content:   "Expired content",
		CreatedAt: time.Now().Add(-48 * time.Hour),
		TTL:       24 * time.Hour,
	}
	saved, err := engine.Store(ctx, expired)
	if err != nil {
		t.Fatalf("failed to store expired record: %v", err)
	}
	if saved == nil {
		t.Fatal("expected saved record")
	}

	_, err = engine.Retrieve(ctx, "expired-id", memory.LayerSession)
	if err == nil {
		t.Fatal("expected expired record retrieval to fail")
	}

	h := handler.NewMemoryHandler(engine, nil)
	req := httptest.NewRequest("GET", "/v1/memory/get?id=expired-id&layer=session", nil)
	w := httptest.NewRecorder()
	h.Get(w, req)
	if w.Code != http.StatusNotFound {
		t.Fatalf("expected 404 for expired record, got %d: %s", w.Code, w.Body.String())
	}
}

File: sdk/typescript/src/__tests__/memory.test.ts
import { describe, it, expect, beforeEach, vi } from 'vitest';
import { AxiosInstance } from 'axios';
import { MemoryAPI } from '../memory';
import { AosClientConfig, MemoryStoreRequest, MemoryStoreResponse, MemorySearchResponse, MemoryGetResponse, MemoryDeleteResponse, MemoryPromoteResponse, MemoryStatsResponse } from '../types';

// =============================================================================
// Helpers
// =============================================================================

function mockHttp() {
  return {
    get: vi.fn(),
    post: vi.fn(),
    put: vi.fn(),
    delete: vi.fn(),
  } as unknown as AxiosInstance;
}

function mockConfig(): Required<AosClientConfig> {
  return {
    baseUrl: 'http://test:1',
    apiKey: '',
    timeout: 5000,
    retryCount: 1,
    headers: {},
  };
}

// =============================================================================
// MemoryAPI
// =============================================================================

describe('MemoryAPI', () => {
  let http: ReturnType<typeof mockHttp>;
  let api: MemoryAPI;

  beforeEach(() => {
    http = mockHttp();
    api = new MemoryAPI(http, mockConfig());
  });

  // ---- store() ----

  describe('store()', () => {
    it('stores a memory record', async () => {
      const request: MemoryStoreRequest = {
        type: 'decision',
        layer: 'session',
        content: 'Decided to use PostgreSQL',
        priority: 1,
        ttl: '24h',
        metadata: { key: 'value' },
      };
      const response: MemoryStoreResponse = { id: 'mem-001', createdAt: '2025-01-01T00:00:00Z' };
      (http.post as any).mockResolvedValue({ data: response });

      const result = await api.store(request);

      expect(result).toEqual(response);
      expect(http.post).toHaveBeenCalledWith('/v1/memory/store', request);
    });

    it('stores with minimal fields', async () => {
      const response: MemoryStoreResponse = { id: 'mem-002', createdAt: '2025-01-01T00:00:00Z' };
      (http.post as any).mockResolvedValue({ data: response });

      const result = await api.store({ content: 'Minimal memory' });

      expect(result).toEqual(response);
      expect(http.post).toHaveBeenCalledWith('/v1/memory/store', { content: 'Minimal memory' });
    });

    it('throws on empty content', async () => {
      await expect(api.store({ content: '' })).rejects.toThrow('Memory content is required');
    });

    it('throws on undefined content', async () => {
      await expect(api.store({ content: undefined as any })).rejects.toThrow('Memory content is required');
    });
  });

  // ---- retrieve() ----

  describe('retrieve()', () => {
    it('retrieves a memory record by ID and layer', async () => {
      const record = { id: 'mem-001', type: 'decision', layer: 'session', content: 'test', priority: 0, createdAt: '2025-01-01T00:00:00Z' };
      (http.get as any).mockResolvedValue({ data: { record } });

      const result = await api.retrieve('mem-001', 'session');

      expect(result).toEqual(record);
      expect(http.get).toHaveBeenCalledWith('/v1/memory/get', { params: { id: 'mem-001', layer: 'session' } });
    });

    it('throws on empty ID', async () => {
      await expect(api.retrieve('', 'session')).rejects.toThrow('Memory record ID is required');
    });

    it('throws on empty layer', async () => {
      await expect(api.retrieve('mem-001', '')).rejects.toThrow('Memory layer is required');
    });
  });

  // ---- search() ----

  describe('search()', () => {
    it('searches memory records', async () => {
      const response: MemorySearchResponse = {
        records: [{ id: '1', type: 'pattern', layer: 'session', content: 'test', priority: 0, createdAt: '2025-01-01T00:00:00Z' }],
        total: 1,
      };
      (http.get as any).mockResolvedValue({ data: response });

      const result = await api.search('test query');

      expect(result).toEqual(response);
      expect(http.get).toHaveBeenCalledWith('/v1/memory/search', { params: { query: 'test query' } });
    });

    it('passes optional filters', async () => {
      (http.get as any).mockResolvedValue({ data: { records: [], total: 0 } });

      await api.search('q', { limit: 5, types: ['pattern'], layers: ['session'] });

      expect(http.get).toHaveBeenCalledWith('/v1/memory/search', {
        params: { query: 'q', limit: 5, types: ['pattern'], layers: ['session'] },
      });
    });

    it('throws on empty query', async () => {
      await expect(api.search('')).rejects.toThrow('Search query is required');
    });

    it('handles empty results', async () => {
      (http.get as any).mockResolvedValue({ data: { records: [], total: 0 } });

      const result = await api.search('nothing');
      expect(result.records).toHaveLength(0);
      expect(result.total).toBe(0);
    });
  });

  // ---- delete() ----

  describe('delete()', () => {
    it('deletes a memory record', async () => {
      (http.delete as any).mockResolvedValue({ data: { success: true } });

      const result = await api.delete('mem-001');

      expect(result).toEqual({ success: true });
      expect(http.delete).toHaveBeenCalledWith('/v1/memory/delete', { params: { id: 'mem-001' } });
    });

    it('deletes with optional layer', async () => {
      (http.delete as any).mockResolvedValue({ data: { success: true } });

      await api.delete('mem-001', 'session');

      expect(http.delete).toHaveBeenCalledWith('/v1/memory/delete', { params: { id: 'mem-001', layer: 'session' } });
    });

    it('throws on empty ID', async () => {
      await expect(api.delete('')).rejects.toThrow('Memory record ID is required');
    });
  });

  // ---- promote() ----

  describe('promote()', () => {
    it('promotes a record between layers', async () => {
      const record = { id: 'mem-001', type: 'decision', layer: 'project', content: 'test', priority: 5, createdAt: '2025-01-01T00:00:00Z' };
      (http.post as any).mockResolvedValue({ data: { record } });

      const result = await api.promote('mem-001', 'session', 'project');

      expect(result).toEqual({ record });
      expect(http.post).toHaveBeenCalledWith('/v1/memory/promote', {
        id: 'mem-001',
        from_layer: 'session',
        to_layer: 'project',
      });
    });

    it('throws on empty ID', async () => {
      await expect(api.promote('', 'session', 'project')).rejects.toThrow('Memory record ID is required');
    });

    it('throws on empty fromLayer', async () => {
      await expect(api.promote('mem-001', '', 'project')).rejects.toThrow('fromLayer is required');
    });

    it('throws on empty toLayer', async () => {
      await expect(api.promote('mem-001', 'session', '')).rejects.toThrow('toLayer is required');
    });
  });

  // ---- stats() ----

  describe('stats()', () => {
    it('retrieves memory engine statistics', async () => {
      const response: MemoryStatsResponse = {
        layers: {
          session: { recordCount: 10, sizeBytes: 1024 },
          project: { recordCount: 5, sizeBytes: 512 },
        },
      };
      (http.get as any).mockResolvedValue({ data: response });

      const result = await api.stats();

      expect(result).toEqual(response);
      expect(http.get).toHaveBeenCalledWith('/v1/memory/stats');
    });
  });

  // ---- error propagation ----

  describe('error propagation', () => {
    it('propagates HTTP errors on store', async () => {
      (http.post as any).mockRejectedValue(new Error('Server Error'));
      await expect(api.store({ content: 'test' })).rejects.toThrow('Server Error');
    });

    it('propagates HTTP errors on retrieve', async () => {
      (http.get as any).mockRejectedValue(new Error('Not Found'));
      await expect(api.retrieve('id', 'layer')).rejects.toThrow('Not Found');
    });

    it('propagates HTTP errors on search', async () => {
      (http.get as any).mockRejectedValue(new Error('Internal Error'));
      await expect(api.search('query')).rejects.toThrow('Internal Error');
    });

    it('propagates HTTP errors on delete', async () => {
      (http.delete as any).mockRejectedValue(new Error('Not Allowed'));
      await expect(api.delete('id')).rejects.toThrow('Not Allowed');
    });

    it('propagates HTTP errors on promote', async () => {
      (http.post as any).mockRejectedValue(new Error('Conflict'));
      await expect(api.promote('id', 'a', 'b')).rejects.toThrow('Conflict');
    });

    it('propagates HTTP errors on stats', async () => {
      (http.get as any).mockRejectedValue(new Error('Unavailable'));
      await expect(api.stats()).rejects.toThrow('Unavailable');
    });
  });
});
